File: server/fastapi/goban_server_fastapi/records/rest/get.py
from datetime import datetime
from typing import Literal, Optional

from pydantic import BaseModel
from sqlalchemy import asc, select

from fastapi import Depends, HTTPException
from goban_server_fastapi.auth import User, jwt_user
from goban_server_fastapi.db import DbClient, dictify
from goban_server_fastapi.records.board import BoardState
from goban_server_fastapi.records.models import Move, Record
from goban_server_fastapi.rest import app


class CoordinateModel(BaseModel):
    x: int
    y: int


class Stone(CoordinateModel):
    color: Literal["B", "W"]


class MoveModel(BaseModel):
    position: Optional[int]
    color: Literal["B", "W"]
    captures: list[CoordinateModel]


class GetResponse(BaseModel):
    id: int
    owner: int
    board_size: Literal[9, 13, 19]
    created: datetime
    name: str
    black_player: str
    white_player: str
    comment: str
    handicap: int
    komi: float
    ruleset: Literal["AGA", "JPN", "CHN"]
    winner: Literal["U", "B", "W"]
    stones: list[Stone]
    moves: list[MoveModel]


@app.get(
    "/api/records/{record_id}/",
    status_code=200,
    response_model=GetResponse,
)
def get_record(
    record_id: int,
    db: DbClient = Depends(),
    current_user: User = Depends(jwt_user),
) -> Optional[Record]:
    with db.session() as session:
        record: Record = session.scalar(
            select(Record)
            .where(Record.id == record_id)
            .where(Record.owner_id == current_user.id)
        )
        if record is None:
            raise HTTPException(status_code=404)
        moves = session.scalars(
            select(Move).where(Move.record_id ==
                               record.id).order_by(asc(Move.move))
        )
        board_state = BoardState(size=record.board_size)
        moves = [
            {
                "position": move.position,
                "color": move.color,
                "captures": [
                    {"x": x, "y": y}
                    for (x, y) in board_state.play_move(
                        move.position % record.board_size,
                        move.position // record.board_size,
                        move.color,
                    )
                ]
                if move.position is not None
                else [],
            }
            for move in moves
        ]
        sorted_stones = sorted(
            board_state.stones.items(), key=lambda x: tuple(reversed(x[0]))
        )
        return {
            **dictify(record),
            "owner": record.owner_id,
            "stones": [
                {"x": x, "y": y, "color": color} for ((x, y), color) in sorted_stones
            ],
            "moves": moves,
        }

File: server/deno/src/record/get.ts
import { register } from "/router.ts";
import { sql } from "/db.ts";
import { getUserId } from "/auth/util.ts";
import { Board, Color, Move } from "/go.ts";

register("GET", "/api/records/{id}/", async (request, { id: recordId }) => {
  const userId = await getUserId(request);
  let rows =
    await sql`SELECT board_size, name, black_player, white_player, comment, handicap, komi, ruleset, winner, created FROM record_record 
  WHERE
  owner_id=${userId} AND id=${recordId}`;
  if (rows.length == 0) {
    return new Response("Not found", { status: 404 });
  }
  const {
    board_size,
    created,
    name,
    black_player,
    white_player,
    comment,
    handicap,
    komi,
    ruleset,
    winner,
  } = rows[0];
  rows =
    await sql`SELECT position, color FROM record_move WHERE record_id=${recordId} ORDER BY move ASC`;
  console.log("mmoves", rows);
  const board = new Board(board_size);
  const moves: Move[] = rows.map((row) => {
    if (row.position === null) {
      return [null, Color.fromString(row.color)];
    }
    return [board.toCoord(row.position), Color.fromString(row.color)];
  });
  const movesAndCaptures = board.playMoves(moves);
  const stones = [];
  for (let pos = 0; pos < board_size * board_size; pos += 1) {
    const color = board.stones[pos];
    if (color !== null) {
      const [x, y] = board.toCoord(pos);
      stones.push({ x, y, color: Color.toString(color) });
    }
  }
  return new Response(
    JSON.stringify({
      id: Number.parseInt(recordId, 10),
      owner: userId,
      board_size,
      created,
      name,
      black_player,
      white_player,
      comment,
      handicap,
      komi,
      ruleset,
      winner,
      moves: movesAndCaptures.map(([[coord, color], captures]) => ({
        position: (coord !== null) ? { x: coord[0], y: coord[1] } : null,
        color: Color.toString(color),
        captures: captures.map(([x, y]) => ({ x, y })),
      })),
      stones,
    }),
  );
});
